dbms/src/Storages/StorageS3.cpp

```diff
@@ -0,0 +1,155 @@
+            return name;
+        }
+
+        Block readImpl() override
+        {
+            return reader->read();
+        }
+
+        Block getHeader() const override
+        {
+            return reader->getHeader();
+        }
+
+        void readPrefixImpl() override
+        {
+            reader->readPrefix();
+        }
+
+        void readSuffixImpl() override
+        {
+            reader->readSuffix();
+        }
+
+    private:
+        String name;
+        std::unique_ptr<ReadBufferFromS3> read_buf;
+        BlockInputStreamPtr reader;
+    };
+
+    class StorageS3BlockOutputStream : public IBlockOutputStream
+    {
+    public:
+        StorageS3BlockOutputStream(const Poco::URI & uri,
+            const String & format,
+            const Block & sample_block_,
+            const Context & context,
+            const ConnectionTimeouts & timeouts)
+            : sample_block(sample_block_)
+        {
+            write_buf = std::make_unique<WriteBufferFromS3>(uri, timeouts);
+            writer = FormatFactory::instance().getOutput(format, *write_buf, sample_block, context);
+        }
+
+        Block getHeader() const override
+        {
+            return sample_block;
+        }
+
+        void write(const Block & block) override
+        {
+            writer->write(block);
+        }
+
+        void writePrefix() override
+        {
+            writer->writePrefix();
+        }
+
+        void writeSuffix() override
+        {
+            writer->writeSuffix();
+            writer->flush();
+            write_buf->finalize();
+        }
+
+    private:
+        Block sample_block;
+        std::unique_ptr<WriteBufferFromS3> write_buf;
+        BlockOutputStreamPtr writer;
+    };
+}
+
+
+BlockInputStreams StorageS3::read(const Names & column_names,
+    const SelectQueryInfo & /*query_info*/,
+    const Context & context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    unsigned /*num_streams*/)
+{
+    BlockInputStreamPtr block_input = std::make_shared<StorageS3BlockInputStream>(uri,
+        format_name,
+        getName(),
+        getHeaderBlock(column_names),
+        context,
+        max_block_size,
+        ConnectionTimeouts::getHTTPTimeouts(context));
+
+
+    auto column_defaults = getColumns().getDefaults();
+    if (column_defaults.empty())
+        return {block_input};
+    return {std::make_shared<AddingDefaultsBlockInputStream>(block_input, column_defaults, context)};
+}
+
+void StorageS3::rename(const String & /*new_path_to_db*/, const String & /*new_database_name*/, const String & /*new_table_name*/) {}
+
+BlockOutputStreamPtr StorageS3::write(const ASTPtr & /*query*/, const Context & /*context*/)
+{
+    return std::make_shared<StorageS3BlockOutputStream>(
+        uri, format_name, getSampleBlock(), context_global, ConnectionTimeouts::getHTTPTimeouts(context_global));
+}
+
+void registerStorageS3(StorageFactory & /*factory*/)
+{
+    // TODO. See #1394?
```

Comment:
You can allow the user to have S3 tables in database - the same as for File, MySQL tables, etc.
It doesn't relate to #1394

Reply:
+